OneMore/Commands/Navigator/NaviPages.cs

```diff
@@ -0,0 +1,219 @@
+            return NPinyin.Pinyin.GetInitials(s, gb2312);
+        }
+
+        /// <summary>
+        /// Indexing the pages by PinYin first letter
+        /// </summary>
+        /// <param name="inXML"></param>
+        /// <param name="strKeyWords"></param>
+        /// <returns></returns>
+        public XElement FilterXml(XElement inXML, string strKeyWords)
+        {
+            string strSecKw, strPageKw;
+            XElement root = inXML;
+            if (strKeyWords.Contains(" "))
+            {
+                strSecKw = strKeyWords.Substring(0, strKeyWords.IndexOf(" "));  //sub-string before SPACE
+                strPageKw = strKeyWords.Substring(strKeyWords.IndexOf(" ") + 1); //sub-string after SPACE
+            }
+            else
+            {
+                strSecKw = strKeyWords;
+                strPageKw = null;
+            }
+
+            string RecycleBinSec = "onenote recyclebin";
+
+            foreach (var NBSec in root.Elements().Elements())
+            {
+                string NBSecName = PinYinFirstLetter(NBSec.Attribute("name").Value).ToLower();
+                if (!NBSecName.Contains(strSecKw) || NBSecName.Contains(RecycleBinSec))
+                    NBSec.RemoveAll();
+
+            }
+            if (strPageKw != null)
+            {
+                foreach (var NBPage in root.Elements().Elements().Elements())
+                {
+                    string NBPageName = PinYinFirstLetter(NBPage.Attribute("name").Value).ToLower();
+                    if (!NBPageName.Contains(strPageKw))
+                        NBPage.RemoveAll();
+                }
+            }
+
+            // delete empty page
+            foreach (XElement child in root.Descendants().Reverse())
+            {
+                if (!child.HasElements && string.IsNullOrEmpty(child.Value) && !child.HasAttributes) child.Remove();
+
+            }
+
+            // delete empty section
+            foreach (XElement child in root.Elements().Elements().Reverse())
+            {
+                if (!child.HasElements) child.Remove();
+
+            }
+
+            //delete empty notebook
+            foreach (var child1 in root.Elements().Reverse())
+            {
+                if (!child1.HasElements) child1.Remove();
+            }
+            return root;
+        }
+
+             
+        private void tBoxKw_TextChanged(object sender, EventArgs e)
+        {
+            using var one = new OneNote();
+            XElement FilteredPageXml;
+            var AllPageXML = one.GetAllPages();
+            try
+            {
+                string strKeyWords = tBoxKw.Text.ToLower();
+                if(strKeyWords=="")
+                {
+                    FilteredPageXml = AllPageXML;
+                }
+                else
+                {
+                    FilteredPageXml = FilterXml(AllPageXML, strKeyWords);
+                }
+
+                hierarchyView11.Nodes.Clear();
```

Comment:
This is a poor use of TreeView, causing flickering and slowness with each keystroke. You also don't need to fetch AllPages on every keystroke since it is unlikely the hierarchy would change while this dialog is open.

Reply:
> This is a poor use of TreeView, causing flickering and slowness with each keystroke. You also don't need to fetch AllPages on every keystroke since it is unlikely the hierarchy would change while this dialog is open.

It seems that this variable **AllPageXML** is passed by reference, not by value.
If the value is not fetched very loop, its value will be changed by the function **FilterXML**.
I've tried to define a class with a copy element, but still no luck here.
As for the flickering, I trid serveral methods, also no luck.
Could you give any comments on how to solve those two problems? Thanks.